Mathlib/Combinatorics/Quiver/Path/Vertices.lean

```diff
@@ -0,0 +1,198 @@
+
+open List
+
+variable {V : Type*} [Quiver V]
+
+/-- The end vertex of a path. A path `p : Path a b` has `p.end = b`. -/
+def «end» {a : V} : ∀ {b : V}, Path a b → V
+  | b, _ => b
+
+@[simp]
+lemma end_cons {a b c : V} (p : Path a b) (e : b ⟶ c) : (p.cons e).end = c := rfl
+
+/-- The list of vertices in a path, including the start and end vertices. -/
+def vertices {a : V} : ∀ {b : V}, Path a b → List V
+  | _, nil => [a]
+  | _, cons p e => (p.vertices).concat (p.cons e).end
+
+@[simp]
+lemma mem_verticesSet_iff {a b : V} (p : Path a b) (v : V) :
+    v ∈ {v | v ∈ p.vertices} ↔ v ∈ p.vertices := Iff.rfl
+
+lemma mem_verticesFinset_iff [DecidableEq V] {a b : V} (p : Path a b) {x : V} :
+    x ∈ p.vertices.toFinset ↔ x ∈ p.vertices :=
+  List.mem_toFinset
+
+lemma mem_initFinset_iff [DecidableEq V] {a b : V} (p : Path a b) {x : V} :
+    x ∈ p.vertices.dropLast.toFinset ↔ x ∈ p.vertices.dropLast :=
+  List.mem_toFinset
+
+@[simp]
+lemma vertices_nil (a : V) : (nil : Path a a).vertices = [a] := rfl
+
+@[simp]
+lemma vertices_cons {a b c : V} (p : Path a b) (e : b ⟶ c) :
+  (p.cons e).vertices = p.vertices.concat c := rfl
+
+lemma mem_verticesSet [DecidableEq V] {a b : V} (p : Path a b) (v : V) :
+    v ∈ p.vertices.toFinset ↔ v ∈ p.vertices := by simp
+
+/-- The vertex list of `cons` — convenient `simp` form. -/
+lemma mem_vertices_cons {a b c : V} (p : Path a b)
+    (e : b ⟶ c) {x : V} :
+    x ∈ (p.cons e).vertices ↔ x ∈ p.vertices ∨ x = c := by
+  simp only [vertices_cons]
+  simp_all only [concat_eq_append, mem_append, mem_cons, not_mem_nil, or_false]
+
+lemma verticesSet_nil {a : V} : {v | v ∈ (nil : Path a a).vertices} = {a} := by
+  simp only [vertices_nil, List.mem_singleton, Set.ext_iff, Set.mem_singleton_iff]
+  exact fun x ↦ Set.mem_setOf
+
+@[simp]
+lemma verticesSet_cons {a b c : V} (p : Path a b) :
+  {v | v ∈ p.vertices ∨ v = c} = {v | v ∈ p.vertices} ∪ {c} := by
+  simp only [Set.union_singleton]
+  ext v
+  simp only [mem_verticesSet_iff, Set.mem_insert_iff]
+  rw [or_comm]; simp
+
+/-- The length of vertices list equals path length plus one -/
+@[simp]
+lemma vertices_length {V : Type*} [Quiver V] {a b : V} (p : Path a b) :
+    p.vertices.length = p.length + 1 := by
+  induction p with
+  | nil => simp only [vertices_nil, List.length_cons, List.length_nil, zero_add, length_nil]
+  | cons p' e ih =>
+    simp only [vertices_cons, length_cons, List.length_concat, ih]
+
+lemma length_vertices_pos {a b : V} (p : Path a b) :
+    0 < p.vertices.length := by simp
+
+lemma vertices_ne_nil {a : V} {b : V} (p : Path a b) : p.vertices ≠ [] := by
+  simp [← List.length_pos_iff_ne_nil]
+
+/-- The head of the vertices list is the start vertex -/
+@[simp]
+lemma vertices_head? {a b : V} (p : Path a b) : p.vertices.head? = some a := by
+  induction p with
+  | nil => simp only [vertices_nil, List.head?_cons]
+  | cons p' e ih => simp [ih]
+
+@[simp]
+lemma length_toPath {a b : V} (e : a ⟶ b) : e.toPath.length = 1 := rfl
```

Comment:
Please move this to the file `Combinatorics.Quiver.Path`, somewhere just below the definition of [Quiver.Path.length](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Combinatorics/Quiver/Path.html#Quiver.Path.length)